components/devtools/actors/network_event.rs

```diff
@@ -499,7 +533,126 @@ impl NetworkEventActor {
         }
     }
 
-    pub fn total_time(&self) -> Duration {
-        self.request.connect_time + self.request.send_time
+    pub fn total_time(request: &HttpRequest) -> Duration {
+        request.connect_time + request.send_time
+    }
+
+    pub fn resource_updates(&self) -> NetworkEventResource {
+        let mut resource_updates = Map::new();
+
+        resource_updates.insert(
+            "requestCookiesAvailable".to_owned(),
+            Value::Bool(self.request_cookies.is_some()),
+        );
+
+        resource_updates.insert(
+            "requestHeadersAvailable".to_owned(),
+            Value::Bool(self.request_headers.is_some()),
+        );
+
+        resource_updates.insert(
+            "responseHeadersAvailable".to_owned(),
+            Value::Bool(self.response_headers.is_some()),
+        );
+        resource_updates.insert(
+            "responseCookiesAvailable".to_owned(),
+            Value::Bool(self.response_cookies.is_some()),
+        );
+        resource_updates.insert(
+            "responseStartAvailable".to_owned(),
+            Value::Bool(self.response_start.is_some()),
+        );
+        resource_updates.insert(
+            "responseContentAvailable".to_owned(),
+            Value::Bool(self.response_content.is_some()),
+        );
+
+        resource_updates.insert(
+            "totalTime".to_string(),
+            Value::from(self.total_time.as_secs_f64()),
+        );
+
+        resource_updates.insert(
+            "securityState".to_string(),
+            Value::String(self.security_state.clone()),
+        );
+
+        if let Some(ref content) = self.response_content {
+            let content_map = serde_json::to_value(content)
+                .unwrap()
+                .as_object()
+                .unwrap()
+                .clone();
+
+            for (key, value) in content_map {
+                resource_updates.insert(key, value);
+            }
```

Comment:
Let's make this a helper function that we can call to reduce code duplication. We can use `<T: Serialize>` for a generic argument.